packages/core/src/libraries/hook/hook-context-manager.ts

```diff
@@ -0,0 +1,43 @@
+import { type DataHookEvent } from '@logto/schemas';
+
+type DataHookContext = {
+  event: DataHookEvent;
+  data?: Record<string, unknown>;
+};
+
+type DataHookMetadata = {
+  userAgent?: string;
+  ip: string;
+};
+
+export class DataHookContextManager {
+  contextArray: DataHookContext[] = [];
+
+  constructor(public metadata: DataHookMetadata) {}
+
+  appendContext({ event, data }: DataHookContext) {
+    const existingContext = this.contextArray.find((ctx) => ctx.event === event);
+
+    // Merge with the existing context if event is the same
+    if (existingContext) {
+      this.contextArray = this.contextArray.map((currentContext) => {
+        if (currentContext.event === event) {
+          return {
+            ...currentContext,
+            data: {
+              ...currentContext.data,
+              ...data,
+            },
+          };
+        }
+
```

Comment:
what if we have multiple context objects for the same event? e.g. multiple roles updated

Reply:
good point. consider remove it .